src/Microsoft.Diagnostics.DebugServices.Implementation/DiagnosticLoggingService.cs

```diff
@@ -0,0 +1,170 @@
+        public void Disable()
+        {
+            CloseLogging();
+            Trace.Listeners.Remove(ListenerName);
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Initializes the diagnostic logging service.  Reads the DOTNET_ENABLED_SOS_LOGGING 
+        /// environment variable to log to console or file.
+        /// </summary>
+        /// <param name="logfile"></param>
+        public static void Initialize(string logfile = null)
+        {
+            try
+            {
+                if (string.IsNullOrWhiteSpace(logfile))
+                {
+                    logfile = Environment.GetEnvironmentVariable("DOTNET_ENABLED_SOS_LOGGING");
+                }
+                if (!string.IsNullOrWhiteSpace(logfile))
+                {
+                    Instance.Enable(logfile == "1" ? null : logfile);
+                }
+            }
+            catch (Exception ex) when ( ex is IOException || ex is NotSupportedException || ex is SecurityException || ex is UnauthorizedAccessException)
+            {
+            }
+        }
+
+        /// <summary>
+        /// Sets the console service and the console file logging control service.
+        /// </summary>
+        /// <param name="consoleService">This is used for to log to the console</param>
+        /// <param name="fileLoggingService">This is used to hook the command console output to write the diagnostic log file.</param>
+        public void SetConsole(IConsoleService consoleService, IConsoleFileLoggingService fileLoggingService = null)
+        {
+            _consoleService = consoleService;
+            _fileLoggingService = fileLoggingService;
+        }
+
+        private void CloseLogging()
+        {
+            if (_writer is not null)
+            { 
+                _fileLoggingService?.RemoveStream(_writer.BaseStream);
+                _writer.Flush();
+                _writer.Close();
+                _writer = null;
+            }
+        }
+
+        class LoggingListener : TraceListener
+        {
+            private readonly DiagnosticLoggingService _diagnosticLoggingService;
+
+            internal LoggingListener(DiagnosticLoggingService diagnosticLoggingService)
+                : base(ListenerName)
+            {
+                _diagnosticLoggingService = diagnosticLoggingService;
+            }
+
+            public override void Close()
+            {
+                _diagnosticLoggingService.CloseLogging();
+                base.Close();
+            }
+
+            public override void Write(string message)
+            {
+                if (_diagnosticLoggingService._writer is not null)
+                {
+                    try
+                    {
+                        _diagnosticLoggingService._writer.Write(message);
+                    }
+                    catch (Exception ex) when (ex is IOException || ex is ObjectDisposedException || ex is NotSupportedException)
+                    {
```

Comment:
Should we output an error message so the user knows that log output isn't working?

Reply:
These public Write/WriteLine APIs implemented here never throw any exceptions.  Same with the IConsoleService Write* API's implemented in the new FileLoggingConsoleService.cs.

Reply:
I made a slight change that at least writes the tracing to the console if the writer throws an exception.